File: 0206_reverse_linked_list/reverse_linked_list.cpp
#include <iostream>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;

    // Constructer
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        reverse(nullptr, head);
        return f_head;
    }
private:
    ListNode* f_head = nullptr;
    void reverse(ListNode *prev, ListNode *ptr){
        if (ptr == nullptr) return;
        if (ptr->next == nullptr){
            f_head = ptr;
            ptr->next = prev;
            return;
        }

        reverse(ptr, ptr->next);
        ptr->next = prev;
    }
};

int main(void){

    Solution sol;
    ListNode *head = nullptr;
    ListNode *ptr;

    // for (int i = 0; i < 5; i++){
    //     ListNode *new_node = new ListNode(i + 1);
        
    //     if (i == 0){
    //         head = new_node;
    //         ptr = head;
    //         continue;
    //     }

    //     ptr->next = new_node;
    //     ptr = new_node;
    // }

    for (auto *ptr = head; ptr != nullptr; ptr = ptr->next){
        cout << ptr->val << " ";
    }
    cout << "\n";

    head = sol.reverseList(head);

    for (auto *ptr = head; ptr != nullptr; ptr = ptr->next){
        cout << ptr->val << " ";
    }
    cout << "\n";
}

File: 206_reverse_linked_list/reverse_linked_list.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
def plist(ptr: Optional[ListNode]):
    while (ptr != None):
        print(ptr.val, end=" ")
        ptr = ptr.next
    print()
        
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        curr, prev = head, None
        
        while curr:
            tmp = curr.next
            curr.next = prev
            prev = curr
            curr = tmp
            
        return prev

list = [1,2,3,4,5]
head = None
tail = None

for i in list:
    new_node = ListNode(i)
    
    if not head: 
        head = new_node
        tail = new_node
        continue
    
    tail.next = new_node
    tail = new_node

del tail

sol = Solution()
ans = sol.reverseList(head)

plist(ans)